contracts/perun-vchannel-lockscript/src/entry.rs

```diff
@@ -0,0 +1,146 @@
+// Import from `core` instead of from `std` since we are in no-std mode
+use core::result::Result;
+
+// Import CKB syscalls and structures
+// https://docs.rs/ckb-std/
+use ckb_std::{
+    ckb_constants::Source,
+    ckb_types::{bytes::Bytes, prelude::*},
+    high_level::{load_cell_lock_hash, load_cell_type, load_script},
+    syscalls::SysError,
+};
+use perun_common::{error::Error, perun_types::ChannelConstants};
+
+// The perun-channel-lockscript (pcls) is used to lock access to interacting with a channel and is attached as lock script
+// to the channel-cell (the cell which uses the perun-channel-type-script (pcts) as its type script).
+// A channel defines two participants, each of which has their own unlock_script_hash (also defined in the ChannelConstants.params.{party_a,party_b}).
+// The pcls allows a transaction to interact with the channel, if at least one input cell is present with:
+// - cell's lock script hash == unlock_script_hash of party_a or
+// - cell's lock script hash == unlock_script_hash of party_b
+// We recommend using the secp256k1_blake160_sighash_all script as unlock script and corresponding payment args for the participants.
+//
+// Note: This means, that each participant needs to use a secp256k1_blake160_sighash_all as input to interact with the channel.
+// This should not be a substantial restriction, since a payment input will likely be used anyway (e.g. for funding or fees).
+
+pub fn main() -> Result<(), Error> {
+    // call normal_mode first return and if it fails, call vc_mode return its value
+    let script = load_script()?;
+
+    let args: Bytes = script.args().unpack();
+
+    // return an error if args is empty
+    if !args.is_empty() {
+        return Err(Error::VCLSWithArgs);
+    }
+
+    // Source::GroupInput would have two cells in case of virtual channel dispute and one cell otherwise
+    // We will use this fact to check which mode we are in (1)Normal mode, (2) VC mode
+    let mut counter = 0;
+    for i in 0.. {
+        match load_cell_type(i, Source::GroupInput) {
+            Ok(Some(script)) => script,
+            Ok(None) => panic!("lockscript not found"),
+            Err(SysError::IndexOutOfBound) => break,
+            Err(err) => return Err(err.into()),
+        };
+        counter += 1;
+    }
+
+    if counter == 1 {
+        return honest_mode(); // normal mode is vc without dispute -> close?
+    } else if counter == 2 {
+        return dispute_mode(); // dispute mode -> merge two cells into one
+    } else {
+        return Err(Error::IndexOutOfBound); // not a legitimae
+    }
+}
+
+pub fn dispute_mode() -> Result<(), Error> {
```

Comment:
I am pretty sure that we NEED to allow the intermediate to perform the merging.
This looks to me like only Alice and Bob can do it?

I think we can just have the lock script that always accepts for now...